### ISSUE TITLE
mypy suggests deprecated code

### ISSUE THREAD

**Issue by clo-vis** [python/mypy#16815]
**To Reproduce**

```python
empty: Set[int] = {}
```

**Expected Behavior**
`error: Name "Set" is not defined  [name-defined]`
and _perhaps_ something like
`note: "Use set[int] if the set is mutable or import Set from collections.abc if the set is immutable: from collections.abc import Set`

**Actual Behavior**
```
error: Name "Set" is not defined  [name-defined]
note: Did you forget to import it from "typing"? (Suggestion: "from typing import Set")
```

typing.Set is deprecated (since 3.9)

**Your Environment**

- Mypy version used: 1.8.0 (compiled: yes)
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.11.6

**Comment by CarliJoy** [comment]
> note: "Use set[int] if the set is mutable or import Set from collections.abc if the set is immutable: from collections.abc import Set

This is quite new to me. Can provide a documentation link that describes this?

**Comment by clo-vis** [comment]
[https://docs.python.org/3.9/library/typing.html#corresponding-to-built-in-types](3.9)
[https://docs.python.org/3.10/library/typing.html#corresponding-to-built-in-types](3.10)
[https://docs.python.org/3.11/library/typing.html#corresponding-to-built-in-types](3.11)
[https://docs.python.org/3.12/library/typing.html#corresponding-to-built-in-types](3.12)

**Comment by CarliJoy** [comment]
But where is written, that `set[T]` is mutable? 
I see that  [collections.abc.Set](https://docs.python.org/3.12/library/collections.abc.html#collections.abc.Set) is immutable  but `set` is mutable as [collections.abc.MutableSet](https://docs.python.org/3.12/library/collections.abc.html#collections.abc.MutableSet) is as well.

That is quite a bit confusing...

I get that `set` is mutable because it is not a `frozenset` but naming it exactly different with `collections.abc`, why?
Given that it is so confusing, the note shown to the users should reflect that.

**Comment by JelleZijlstra** [comment]
Yes, this is confusing. We have a special case in flake8-pyi to deal with it: search for Y025 in https://github.com/PyCQA/flake8-pyi/blob/main/ERRORCODES.md.

It's a little off topic for this issue though: I assume mypy suggests the same deprecated code if you write `List[int]` without importing `List`.

I think it's still appropriate for mypy to suggest `from typing import ...` as long as we still support a version where `list[int]` doesn't work (i.e., Python 3.8). We could suggest `list[int]` instead when we know it's safe (i.e., when the project supports only 3.9+).

**Comment by hauntsaninja** [comment]
(In case its any consolation, note that while these symbols are deprecated, CPython will not remove them for a long long time, if ever)

**Comment by clo-vis** [comment]
Description updated (to show the problem also with `Iterable`)
